infinite loop in compact

the attached database contains just 1 empty table.
[test.mv.db.zip](https://github.com/user-attachments/files/16200911/test.mv.db.zip)

opening and closing it with h2shell leads to infinite compaction until it times out.

of course this also happens with lots of tables and data and when embedded, 
but this is a minimised example.

i inserted a few writes into the code to illustrate the loop, see the attached log.
[test.log](https://github.com/user-attachments/files/16200912/test.log)

when i open and immediatelay close this db with the 2.3.230 shell, after 3 and 5 and 8 times (yes, really) i get exceptions like:

```
sql> quit
SQL Exception: General error: "General error: ""org.h2.mvstore.MVStoreException: java.lang.AssertionError: 28672 != 20480 [2.2.224/3]"" [50000-224]" [50000-224]
org.h2.jdbc.JdbcSQLNonTransientException: General error: "General error: ""org.h2.mvstore.MVStoreException: java.lang.AssertionError: 28672 != 20480 [2.2.224/3]"" [50000-224]" [50000-224]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:566)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:212)
        at org.h2.engine.Database.throwLastBackgroundException(Database.java:1817)
        at org.h2.engine.SessionLocal.close(SessionLocal.java:890)
        at org.h2.jdbc.JdbcConnection.close(JdbcConnection.java:391)
        at org.h2.tools.Shell.promptLoop(Shell.java:312)
        at org.h2.tools.Shell.runTool(Shell.java:150)
        at org.h2.tools.Shell.main(Shell.java:80)
Caused by: org.h2.message.DbException: General error: "org.h2.mvstore.MVStoreException: java.lang.AssertionError: 28672 != 20480 [2.2.224/3]" [50000-224]
        at org.h2.message.DbException.get(DbException.java:212)
        at org.h2.message.DbException.convert(DbException.java:407)
        at org.h2.mvstore.db.Store.lambda$new$0(Store.java:122)
        at org.h2.mvstore.MVStore.handleException(MVStore.java:1546)
        at org.h2.mvstore.MVStore.panic(MVStore.java:371)
        at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1529)
        at org.h2.mvstore.FileStore.lambda$serializeAndStore$3(FileStore.java:1432)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.h2.jdbc.JdbcSQLNonTransientException: General error: "org.h2.mvstore.MVStoreException: java.lang.AssertionError: 28672 != 20480 [2.2.224/3]" [50000-224]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:566)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        ... 12 more
Caused by: org.h2.mvstore.MVStoreException: java.lang.AssertionError: 28672 != 20480 [2.2.224/3]
        at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
        ... 7 more
Caused by: java.lang.AssertionError: 28672 != 20480
        at org.h2.mvstore.RandomAccessStore.shrinkStoreIfPossible(RandomAccessStore.java:679)
        at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:377)
        at org.h2.mvstore.RandomAccessStore.writeChunk(RandomAccessStore.java:28)
        at org.h2.mvstore.FileStore.storeBuffer(FileStore.java:1524)
        ... 6 more

```

--- a/h2/src/main/org/h2/mvstore/FileStore.java
+++ b/h2/src/main/org/h2/mvstore/FileStore.java
@@ -693,7 +693,9 @@ public abstract class FileStore<C extends Chunk<C>>
     }
 
     private boolean isRewritable(C chunk, long time) {
-        return chunk.isRewritable() && isSeasonedChunk(chunk, time);
+        return chunk.isRewritable() && isSeasonedChunk(chunk, time)
+                // to prevent last saved chunk from being re-written as it may cause "endless" re-write loop
+                && chunk.version < getMvStore().getCurrentVersion() - 1;
     }
 
     /**

--- a/h2/src/main/org/h2/mvstore/MVStoreTool.java
+++ b/h2/src/main/org/h2/mvstore/MVStoreTool.java
@@ -153,6 +153,7 @@ public class MVStoreTool {
                 Chunk c;
                 try {
                     c = new SFChunk(Chunk.readChunkHeader(buffer));
+                    c.block = pos / blockSize;
                 } catch (MVStoreException e) {
                     // Chunks are not always contiguous (due to chunk compaction/move/drop and space re-use)
                     // Blocks following a chunk can therefore contain something else than a valid chunk header

--- a/h2/src/main/org/h2/mvstore/RandomAccessStore.java
+++ b/h2/src/main/org/h2/mvstore/RandomAccessStore.java
@@ -657,11 +657,11 @@ public abstract class RandomAccessStore extends FileStore<SFChunk>
         assert reservedAreaHigh > 0 || block <= chunk.block : block + " " + chunk;
         ByteBuffer readBuff = readFully(chunk, start, length);
         writeFully(null, pos, readBuff);
-        free(start, length);
-        // can not set chunk's new block/len until it's fully written at new location,
+        // can not set chunk's new block until it's fully written at new location,
         // because concurrent reader can pick it up prematurely,
         chunk.block = block;
         chunk.next = 0;
+        free(start, length);
         saveChunkMetadataChanges(chunk);
         return true;
     }

